astropy/coordinates/builtin_frames/intermediate_rotation_transforms.py

```diff
@@ -90,9 +90,10 @@ def gcrs_precession_mat(equinox):
 @frame_transform_graph.transform(FunctionTransformWithFiniteDifference, GCRS, TETE)
 def gcrs_to_tete(gcrs_coo, tete_frame):
     # first apply GCRS self transform to correct time and observatory position/velocity
+    obsgeoloc, obsgeovel = tete_frame.location.get_gcrs_posvel(tete_frame.obstime)
```

Comment:
This is definitely not the way to go here. You are calculating the BPN matrix twice. This should be done in the same manner as `gcrs_to_cirs()`.

Reply:
This is why I liked @StuartLittlefair's `EarthLocation.get_posvel()` method.

Reply:
OK, that makes sense. I'll adjust.  There is also a real failure in comparison with horizons, so I may be missing more than this...

Reply:
Just to be sure, is there still a benefit from having the GCRS<->TETE conversion explicitly defined?

Reply:
(On `get_posvel()`: only useful if we stick with `obsgeoloc, obsgeovel` for GCRS...)

Reply:
Never mind on the benefit - things break greatly if I just remove the transformations...

Reply:
How else would we get the apparent position of a solar system body? Those are returned in the GCRS frame. For that matter, as far as I can see any ICRS<->TETE goes through GCRS.

Reply:
I do think that you should stick with `obseoloc, obsgeovel` for GCRS as GCRS is the natural Earth centered frame for a non-terrestrial observer location.

Reply:
Also, `obsgeloc` is definitely needed in order to calculate the time delay to the observer in `get_body()`.

Reply:
This is why I advocated for `get_gcrs_posvel()` to be incorporated into `pos_vel()`. That single `EarthLocation` method could handle all of these cases.

Reply:
I suppose another possibility would be to add an optional argument to `get_gcrs_posvel(rbpn=None)` then if it is called without and BPN matrix, it would act just as it does currently, but if it was given one it would use it.

Reply:
OK, I think I got it in a more logical state now. @mkbrewer - Would you mind having another look?

Reply:
I agree with @mkbrewer that we should have something as flexible as `obsgeoloc{vel}` for GCRS. However, we may want to revisit the exact way these are stored. These choices pre-date coordinates having velocities, for example. 

We might want to use a different class to store the observer position, in which case it makes sense for a `pos_vel` routine on `EarthLocation` to return that same class. So having the method there makes sense eventually, but it’s not yet clear it should return a pair of `CartesianRepresentation`s.

Reply:
On `obsgeopos/vel`, I think it is mostly a question of whether it should instead be a `GCRS` instance that holds the position/velocity. Anyway, that is for #10996!

Reply:
p.s. Note that what I'm doing in the new commit is not really quite equivalent to `get_gcrs_posvel(rbpn=None)` -- am I still missing the main speed-up?

Reply:
Well, it basically is. It's just that it requires another function whereas all of this could be incorporated into a single EarthLocation.get_gcrs_posvel(rpbn=None) method. Anyway, this looks good for now.

Reply:
The main way to speed this up would be to chain matrices together; since we know this is a transform at the same obstime we are just doing rotations to get from ITRS->GCRS so you can avoid some of the flexibility of `transform_to`. 

You could pass a list - in the correct order - of already calculated matrices to transform ITRS into GCRS like so:

```python
matrix_chain = [matrix_transpose(tete_to_itrs_mat(frame.obstime)), rbpn]
loc_gcrs = get_location_gcrs(frame, matrix_chain)
```

Reply:
In fact, that could just be 
```
itrs_to_gcrs = matrix_transpose(tete_to_itrs_mat(frame.obstime) @ rbpn)
loc_gcrs = get_location_gcrs(frame, itrs_to_gcrs))
```
with indeed a simpler `get_location_gcrs`. (I may have swapped things...)